File: python/437. 路径总和 III.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        def count(node:TreeNode,sum:int) -> int:
            if not node:
                return 0
            isMe = 1 if node.val == sum else 0
            leftCount = count(node.left,sum-node.val)
            rightCount = count(node.right,sum-node.val)
            return isMe + leftCount + rightCount

        if not root:
            return 0

        return count(root,sum) + self.pathSum(root.left,sum) + self.pathSum(root.right,sum)

File: go/general/437. 路径总和 III.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func count(node *TreeNode, sum int) int {
	if node == nil {
		return 0
	}
	isMe := 0
	if node.Val == sum {
		isMe = 1
	}
	leftCount := count(node.Left, sum-node.Val)
	rightCount := count(node.Right, sum-node.Val)
	return isMe + leftCount + rightCount
}

func pathSum(root *TreeNode, sum int) int {
	if root == nil {
		return 0
	}
	return count(root, sum) + pathSum(root.Left, sum) + pathSum(root.Right, sum)
}
